### PR TITLE
fix(Storage): Fix the reverse order problem of barrier_manager's sst

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

In rw, SSTs need to be stored in the reverse order of epoch in version.
After https://github.com/risingwavelabs/risingwave/pull/4442.  If we can't sync all cn, we will save the successfully synced SSTs in barrier_manager in positive order of epoch, and commit all SSTs with a checkpoint. So in this version, the order of SSTs is wrong.

It can lead to https://github.com/risingwavelabs/risingwave/issues/4861
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/meta/src/barrier/mod.rs [hunk 1]
```diff
@@ -214,7 +214,7 @@ type CheckpointPost<S> = (
 struct UncommittedMessages<S: MetaStore> {
     uncommitted_checkpoint_post: VecDeque<CheckpointPost<S>>,
     /// Ssts that need to commit with next checkpoint.
-    uncommitted_ssts: Vec<LocalSstableInfo>,
+    uncommitted_ssts: VecDeque<LocalSstableInfo>,
```
Comment (hzxa21) [inline src/meta/src/barrier/mod.rs:217]: Please also add documentation here about what order we preserve and why the ordering matters.

File: src/storage/src/hummock/local_version.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ pub struct LocalVersion {
     pub version_ids_in_use: BTreeSet<HummockVersionId>,
     // TODO: save uncommitted data that needs to be flushed to disk.
     /// Save uncommitted data that needs to be synced or finished syncing.
-    pub sync_uncommitted_data: Vec<(Vec<HummockEpoch>, SyncUncommittedData)>,
+    pub sync_uncommitted_data: VecDeque<(Vec<HummockEpoch>, SyncUncommittedData)>,
```
Comment (hzxa21) [inline src/storage/src/hummock/local_version.rs:47]: Please also add documentation here about what order we preserve and why the ordering matters.

File: src/storage/src/hummock/local_version.rs [hunk 3]
```diff
@@ -170,26 +170,26 @@ impl LocalVersion {
 
     pub fn add_sync_state(
         &mut self,
-        sync_epoch: Vec<HummockEpoch>,
+        sync_epochs: Vec<HummockEpoch>,
         sync_uncommitted_data: SyncUncommittedData,
```
Comment (hzxa21) [inline src/storage/src/hummock/local_version.rs:174]: `sync_epoch` and `sync_uncommitted_data` shoud order their items in descending order of epochs. Let's add a debug assert here.

**Review by wenym1 — APPROVED** [review]
Rest LGTM

File: src/meta/src/barrier/mod.rs (hunk 1 above)
Comment (BugenZhao) [reply src/meta/src/barrier/mod.rs:217]: +1

**Review by hzxa21 — APPROVED** [review]

**Comment by hzxa21** [comment]
Since it is a potential cause to #4861, I will merge this PR first. Please add comments in #4861 to document how this bug happens (you can provide the sequence of steps that can lead to this bug via an exmaple) afterwards. @xxhZs

**Pull request merged**